src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentFactory.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(LSPDocumentFactory))]
+    internal class DefaultLSPDocumentFactory : LSPDocumentFactory
+    {
+        private readonly FileUriProvider _fileUriProvider;
+        private readonly IEnumerable<VirtualDocumentFactory> _virtualDocumentFactories;
+
+        [ImportingConstructor]
+        public DefaultLSPDocumentFactory(
+            FileUriProvider fileUriProvider,
+            [ImportMany] IEnumerable<VirtualDocumentFactory> virtualBufferFactories)
+        {
+            if (fileUriProvider is null)
+            {
+                throw new ArgumentNullException(nameof(fileUriProvider));
+            }
+
+            if (virtualBufferFactories is null)
+            {
+                throw new ArgumentNullException(nameof(virtualBufferFactories));
+            }
+
+            _fileUriProvider = fileUriProvider;
+            _virtualDocumentFactories = virtualBufferFactories;
+        }
+
+        public override LSPDocument Create(ITextBuffer buffer)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            var uri = _fileUriProvider.GetOrCreate(buffer);
+            var virtualDocuments = CreateVirtualDocuments(buffer);
+            var lspDocument = new DefaultLSPDocument(uri, virtualDocuments);
+
+            return lspDocument;
+        }
+
+        private IReadOnlyList<VirtualDocument> CreateVirtualDocuments(ITextBuffer hostDocumentBuffer)
+        {
+            var virtualDocuments = new List<VirtualDocument>();
+            foreach (var factory in _virtualDocumentFactories)
+            {
+                if (factory.TryCreateFor(hostDocumentBuffer, out var virtualDocument))
```

Comment:
Is there any concern about this broadly creating documents for content you don't work with?  E.g. if we add our own CssVirtualDocumentFactory, or maybe there's a CppVirtualDocumentFactory... or should each factory's implementation of `TryCreateFor` be determining when the factory should run based on the `hostDocumentBuffer`?

Currently it looks like (from #1605) you're having the factories check the content type.  Is that going to be extensible - e.g. if I want to use a similar model for PHP support, how will the HtmlVirtualDocumentFactory know to accept that content type?

*Or*, is this a private design that each top level language that wants to follow this needs to implement its own hierarchy of `*VirtualDocumentFactory` types for each language it supports?

Reply:
The goal is to have each language provide their own document factory because that's the piece that creates the "Default" LSPDocument. Each language may need to interoperate with their "Default" lsp document differently so having that type specific to each language made more sense.